source/h2_connection.c

```diff
@@ -941,6 +942,112 @@ static void s_stream_complete(struct aws_h2_connection *connection, struct aws_h
+        /* No pseudoheader detected, we set them from the request */
+        /* Set pseudo headers */
+        struct aws_byte_cursor method;
+        if (aws_http_message_get_request_method(request, &method)) {
+            /* error will happen when the request is invalid */
+            goto error;
+        }
+        if (aws_http_headers_add(result, *aws_h2_pseudoheader_name_to_cursor[PSEUDOHEADER_METHOD], method)) {
+            goto error;
+        }
+        /* we set a default value, "https", for now */
+        struct aws_byte_cursor scheme_cursor = AWS_BYTE_CUR_INIT_FROM_STRING_LITERAL("https");
+        if (aws_http_headers_add(result, *aws_h2_pseudoheader_name_to_cursor[PSEUDOHEADER_SCHEME], scheme_cursor)) {
+            goto error;
+        }
+        /* Set an empty authority for now, if host header field is found, we set it as the value of host */
+        struct aws_byte_cursor authority_cursor;
+        struct aws_byte_cursor host_cursor = AWS_BYTE_CUR_INIT_FROM_STRING_LITERAL("host");
+        if (!aws_http_headers_get(old_headers, host_cursor, &authority_cursor)) {
+            if (aws_http_headers_add(
+                    result, *aws_h2_pseudoheader_name_to_cursor[PSEUDOHEADER_AUTHORITY], authority_cursor)) {
+                goto error;
+            }
+        }
+        struct aws_byte_cursor path_cursor;
+        if (aws_http_message_get_request_path(request, &path_cursor)) {
+            goto error;
+        }
+        if (aws_http_headers_add(result, *aws_h2_pseudoheader_name_to_cursor[PSEUDOHEADER_PATH], path_cursor)) {
+            goto error;
+        }
+    }
+    /* if pseudoheader is included in message, we just convert all the headers from old_headers to result */
+    if (aws_byte_buf_init(&lower_name_buf, alloc, 256)) {
+        goto error;
+    }
+    for (size_t iter = 0; iter < aws_http_headers_count(old_headers); iter++) {
+        /* name should be converted to lower case */
+        if (aws_http_headers_get_index(old_headers, iter, &header_iter)) {
+            goto error;
+        }
+        /* append lower case name to the buffer */
+        aws_byte_buf_append_with_lookup(&lower_name_buf, &header_iter.name, aws_lookup_table_to_lower_get());
+        struct aws_byte_cursor lower_name_cursor = aws_byte_cursor_from_buf(&lower_name_buf);
+        enum aws_http_header_name name_enum = aws_http_lowercase_str_to_header_name(lower_name_cursor);
+        switch (name_enum) {
+            case AWS_HTTP_HEADER_COOKIE:
+                /* split cookie if USE CACHE */
+                if (header_iter.compression == AWS_HTTP_HEADER_COMPRESSION_USE_CACHE) {
+                    struct aws_array_list cookie_chunks;
+                    if (aws_array_list_init_dynamic(&cookie_chunks, alloc, 4, sizeof(struct aws_byte_cursor))) {
+                        goto error;
+                    }
+                    if (aws_byte_cursor_split_on_char(&header_iter.value, ';', &cookie_chunks)) {
```

Comment:
`aws_byte_cursor_next_split()` would let you iterate the cookies with 0 allocations.
```
struct aws_byte_cursor cookie_chunk = {0};
while (aws_byte_cursor_next_split(&header_iter.value, ';', &cookie_chunk)) {
    ...
}
```

why do both these split functions exist? the allocating one came first, the zero-allocation one came second